LateNight: Shouldn't use widget stacks for show/hide operations

ywwg opened the pull request:
This is regarding this bug: https://bugs.launchpad.net/mixxx/+bug/1451187

On first launch, if the user deselects "mixer", all of the sampler / mic / effect racks shouldn't suddenly appear.  

This PR doesn't solve the larger problem of the panels being inconsistent between sizeaware sizes, but it at least establishes that all the panels start off.

I think it works ok but I'd like others to test it to make sure I didn't break anything else.

ywwg committed:
LateNight: Shouldn't use widget stacks for show/hide operations

daschuer commented:
The over all experience is OK, but the view menu does not work if the mixer is shown.

daschuer commented:
Did you have a look to the original issue, the bug in the widget stack. 
I think the widget stack must not change the visible state during initialization, if it has widgets with a persistence visible state.

ywwg committed:
Add unit tests for widget stack

ywwg commented:
Added tests for widget stacks.  The widget stack code works, the problem was that my stacks were badly set up in the skin design.

ywwg commented:
ahhhh I finally have a test that proves why the old skin layout was broken:  the first page of the stack was the "shown" state, but the current page control defaulted to zero so on startup it would show the first page.  This is correct behavior.

ywwg committed:
Added a test which shows how the controls interact on startup if they
disagree

ywwg commented:
added a test for that situation.

ywwg commented:
look at me, I'm working in the wrong branch!

ywwg committed:
more comments to explain

ywwg commented:
> The over all experience is OK, but the view menu does not work if the mixer is shown.

More accurately, the view menu only works when the lower half of the mixxx window is large enough.  The smaller views use different objects to show/hide the panels.

ywwg committed:
LateNight: Hidden WidgetStacks shouldn't be able to affect state.

If a widgetstack is hidden, changing the various trigger COs no longer
cause changes.  This allows LateNight to use the same trigger COs for
all three of its view sizes.  Previously enabling Effects would cause
one of the hidden widgetstacks to disable Samplers, for example.

New unit test for hide / show behavior

ywwg commented:
This appears to fix the consistency issues between window sizes.

daschuer commented on src/test/wwidgetstack_test.cpp:
```diff
@@ -0,0 +1,207 @@
+}
+
+TEST_F(WWidgetStackTest, OnHideBehavior) {
+    // When hiding a page, if the on_hide_select value is -1, the next
```
on_hide_select: Do we need a test for this?

ywwg replied:
this is that test. As the comment says, the on_hide_select values are set in SetUp

daschuer replied:
It is not a test for the usage of m_hideMap wit a dedicated value. 
But it is OK for me to have no test for this. 
IMHO we should go ahead remove this test entirely because, because it fixes an undesired behavior, no one should rely on.

daschuer replied:
This test has probably some flaky nature according to the test slides.

ywwg replied:
> m_hideMap wit a dedicated value.

yes it does, setting hide on page 0 selects page 2 instead of page 1.  I will add a comment.

ywwg replied:
> test slides.

I don't know what this means

ywwg replied:
> flaky

I don't see how it could be flaky.  If I do ./mixxx-test --gtest_repeat=1000 it does not fail.

daschuer replied:
>>   test slides.

> I don't know what this means

The slide show about tests, you have provided here: https://github.com/mixxxdj/mixxx/commit/65048eccc7c67f5aa95ee8b80b2423f5e433a2a3#commitcomment-11106918

daschuer replied:
>>   flaky

> I don't see how it could be flaky. If I do ./mixxx-test --gtest_repeat=1000 it does not fail.

That is not the problem here. It is flaky, because it fixes an undesired behavior. Once the TODO it is fixed this test will fail.

ywwg replied:
that's not what flaky means.  Flaky means that there is a random element in a test that causes it to occasionally fail.  For instance the rubberband test was flaky because it worked on my version but failed with other versions of the library.  Since we don't have direct control over the library version it was a flaky test.

Yes, when the TODO is fixed the test will fail, so the developer will update the test to confirm the new behavior.  That is how it should work.

daschuer commented on src/widget/wwidgetstack.cpp:
```diff
@@ -85,13 +96,32 @@ void WWidgetStack::hideIndex(int index) {
+        return;
+    }
     if (v > 0.0) {
         setCurrentIndex((currentIndex() + 1) % count());
```
This assumes that we have consecutive number of pages. The code above in WWidgetStack::showEvent allows gaps. 
Both versions works for me, but it should be consistent.

ywwg replied:
The pages are consecutive, but having a trigger control per-page is not necessarily true.  So setting a current index this way is safe

daschuer commented on src/widget/wwidgetstack.cpp:
```diff
@@ -85,13 +96,32 @@ void WWidgetStack::hideIndex(int index) {
+    // Set the page triggers to match the current index.
+    for (QMap<int, ControlObject*>::iterator it = m_triggers.begin();
+            it != m_triggers.end(); ++it) {
+        it.value()->set(it.key() == index ? 1.0 : 0.0);
```
Is this loop redundant, we should get et lest a show event for the current widget right? 
But anyway, just check how it is and leaf a comment about the signals we expect.

ywwg replied:
no, because each page may have a trigger CO (like a button for each page) that needs to be consistent with the show/hide state of the stack.  This code sets those trigger COs if they exist.  If the CO is incorrectly on, and the stack is shown, we need to set the CO to off.

This fixes the case where the user has samplers and microphones on and then switches to a stack where only one of them can be shown -- this code makes sure that only one of those COs is on, the one for the page that is shown.

daschuer replied:
Mm I think there is missing a guard somehow. 
I have placed debug messages inside the showEvent  loop in showIndex and hideIndex and that prints the index and the function. 

I got this when fireing up Mixer: 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 0 WWidgetStack::showEvent 
Debug [Main]: 0 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 2 WWidgetStack::showEvent 
Debug [Main]: 2 WWidgetStack::showIndex 
Debug [Main]: 2 WWidgetStack::showIndex 
Debug [Main]: 3 WWidgetStack::showEvent 
Debug [Main]: 3 WWidgetStack::hideIndex 
Debug [Main]: 3 WWidgetStack::hideIndex 
Debug [Main]: Running Mixxx 

This is printed when enabling the Mixer

Debug [Main]: 3 WWidgetStack::showIndex 
Debug [Main]: 3 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showIndex 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 2 WWidgetStack::showEvent 
Debug [Main]: 2 WWidgetStack::showIndex 
Debug [Main]: 2 WWidgetStack::showIndex 
Debug [Main]: 3 WWidgetStack::showEvent 
Debug [Main]: 3 WWidgetStack::hideIndex 
Debug [Main]: 3 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::showEvent 
Debug [Main]: 1 WWidgetStack::hideIndex 
Debug [Main]: 1 WWidgetStack::hideIndex

daschuer commented on src/widget/wwidgetstack.h:
```diff
@@ -81,6 +83,8 @@ class WWidgetStack : public QStackedWidget, public WBaseWidget {
     // signal.
     QMap<int, int> m_hideMap;
+    // A map of the individual page triggers so we can rectify state if needed.
+    QMap<int, ControlObject*> m_triggers;
```
This can be a list.

ywwg replied:
no, because not every page may have a trigger.

daschuer commented:
Looking at the default hide behavior, It works somehow unexpected to me. 

IMHO the default hide index should be the old, not the next. 
If I have selected an index, lets say samples and I need to have a look at effects I toggle the effects button. If I toogle the effects again, you might expect to see nothing as implemented by LateNight, but if this attribute is not set, you may want the old view, but I see no use case to see the next view. 
We have already the next control for this.

ywwg commented:
I agree that the default behavior is counterintuitive, but I want to preserve existing functionality for now.  I will leave a todo to revisit the issue.

ywwg committed:
Add a comment about the widgetstack hide behavior being weird by default

ywwg committed:
WidgetStack: Add a guard around controller-triggered changes while we are showing

ywwg committed:
there is no initial hide event, so respondtochanges should start off

rryan commented:
Hm, show-microphone view menu button doesn't seem to do anything? 
Also Maximize Library checkbox is inverted.

rryan commented on res/skins/LateNight/fx_samplers_container_full.xml:
```diff
@@ -4,61 +4,45 @@
-                    </WidgetGroup>
-                </Children>
-            </WidgetStack>
+          <WidgetGroup trigger="">
```
since it's not in a stack trigger isn't needed, right?

rryan commented on src/widget/wwidgetstack.cpp:
```diff
@@ -114,6 +156,7 @@ void WWidgetStack::addWidgetWithControl(QWidget* pWidget, ControlObject* pContro
                                         int on_hide_select) {
     int index = addWidget(pWidget);
     if (pControl) {
+        m_triggers[index] = pControl;
```
pControl may not live as long as the widgetstack so you should make a COSlave of its configkey -- or make m_triggers a map of the WidgetStackControlListeners and get access to the COThread member variable they have

rryan commented on src/widget/wwidgetstack.cpp:
```diff
@@ -74,24 +75,59 @@ QSize WWidgetStack::minimumSizeHint() const {
+        it.value()->set(it.key() == index ? 1.0 : 0.0);
+    }
+
+    m_bRespondToChanges = true;
```
I'm not sure if respond-to-changes exists because you wanted to ignore trigger CO reactions caused by the above loop -- but if you set the COThread that is the member variable of the WidgetStackControlListener instead of setting the CO then you inherently won't get any trigger reactions since the sets will not trigger valueChanged signals on the COThreads.

ywwg replied:
respond-to-changes exists so that two widgetstacks can share the same trigger CO (show_samplers) and not get in to an argument about who is in charge.  It is a proxy for isvisible.  With the COThread change I can just say "isvisible"

rryan commented on src/widget/wwidgetstack.cpp:
```diff
@@ -131,6 +174,8 @@ void WWidgetStack::addWidgetWithControl(QWidget* pWidget, ControlObject* pContro
     }
 
     if (m_currentPageControl.get() == index) {
+        // The value in the curren page control overrides whatever initial
```
nit: current

ywwg commented:
> Hm, show-microphone view menu button doesn't seem to do anything? 
> Also Maximize Library checkbox is inverted.

these are separate issues from the general widgestack problem.  They'll be addressed in their own PR

rryan commented:
> these are separate issues from the general widgestack problem. They'll be addressed in their own PR

ok -- wasn't sure

ywwg committed:
WWidgetStack: set trigger value through the listeners

ywwg commented:
How do the listener pointers get deleted?   With my updated code I could call qdeleteall on m_listeners.

ywwg commented:
> ok -- wasn't sure

yeah those will just be simple CO name fixes after this goes in.

rryan commented:
> How do the listener pointers get deleted? With my updated code I could call qdeleteall on m_listeners.

You can delete them if you want but they are parented to the WidgetStack so they get deleted anyway.

rryan commented:
lgtm, thanks for adding the test!
